samples/nrf9160/http_application_update/src/main.c

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <stdint.h>
+#include <zephyr.h>
+#include <flash.h>
+#include <bsd.h>
+#include <nrf_socket.h>
+#include <download_client.h>
+#include <logging/log.h>
+#include <gpio.h>
+#include <flash_map.h>
+#include <dfu/mcuboot.h>
+
+#define LED_PORT	LED0_GPIO_CONTROLLER
+
+static struct	device	*flash_dev;
+static bool		is_flash_page_erased[FLASH_PAGE_MAX_CNT];
+static bool		start_dfu;
+static struct	gpio_callback gpio_cb;
+
+static int app_download_client_event_handler(struct download_client *const dfu,
+			enum download_client_evt event, u32_t status);
+
+
+static struct download_client dfu = {
+	.host = CONFIG_DOWNLOAD_HOST,
+	.resource = CONFIG_DOWNLOAD_FILE,
+	.callback = app_download_client_event_handler
+};
+
+
+/**@brief Recoverable BSD library error. */
+void bsd_recoverable_error_handler(uint32_t err)
+{
+	printk("bsdlib recoverable error: %lu\n", err);
+}
+
+/**@brief Irrecoverable BSD library error. */
+void bsd_irrecoverable_error_handler(uint32_t err)
+{
+	while (true) {
+
+	}
+}
+
+
+/**@brief Initialize application. */
+static void app_dfu_init(void)
+{
+	int i;
+
+	for (i = 0; i < FLASH_PAGE_MAX_CNT; i++) {
+		is_flash_page_erased[i] = false;
+	}
+	flash_dev = device_get_binding(DT_FLASH_DEV_NAME);
+	if (flash_dev == 0) {
+		printk("Nordic nRF flash driver was not found!\n");
+		return;
+	}
+
+	int retval = download_client_init(&dfu);
+
+	if (retval != 0) {
+		printk("download_client_init() failed, err %d", retval);
+		return;
+	}
+}
+
+
+/**@brief Start transfer of the file. */
+static void app_dfu_transfer_start(void)
+{
+	int retval = download_client_connect(&dfu);
+
+	if (retval != 0) {
+		printk("download_client_connect() failed, err %d",
+			retval);
+		return;
+	}
+
+	retval = download_client_start(&dfu);
+	if (retval != 0) {
+		printk("download_client_start() failed, err %d",
+			retval);
+		return;
+	}
+}
+
+
+static int app_download_client_event_handler(
+	struct download_client *const dfu,
+	enum download_client_evt event, u32_t error)
+{
+	int err;
+
+	switch (event) {
+	case DOWNLOAD_CLIENT_EVT_DOWNLOAD_FRAG: {
+		struct flash_pages_info info;
+		static u32_t address = CONFIG_FLASH_WRITE_OFFSET;
```

Comment:
This variable is never reset to its default value. Is the address for the second transfer correct? Not sure where you want to write it.

Reply:
`address` renamed to `flash_address` and moved to top of file. Initialized in `app_dfu_init`.